### DISCUSSION TITLE
gatsby build forces NODE_ENV=production regardless of externally setting

### DISCUSSION THREAD

**Discussion by jonmadison** [gatsbyjs/gatsby#31802] [Help] [answered]
Well perhaps not force it during development. Just set it unless the user set something else. We have to have `production` during builds for a number of tools to work correctly.

_Originally posted by @KyleAMathews in https://github.com/gatsbyjs/gatsby/issues/3896#issuecomment-364174486_

it seems like you do force it in gatsby build. trying to do a development build (in order to build my elasticsearch index locally and take advantage of pointing to my local elastic search URL via `.env.development`, and noticing that regardless of setting `NODE_ENV=development` externally before gatsby build, it still forces `NODE_ENV=production`.   

in my gatsby-config:
```
require("dotenv").config({
  path: `.env.${process.env.NODE_ENV}`,
});
console.log(`GATSBY CONFIG\n\n\n\nprocess.env.NODE_ENV:${process.env.NODE_ENV}\n\n\n\n\n`);
process.exit(0)
```

after running `NODE_ENV=development gatsby build`

```
➜  DDG-Illustration-Gatsby-POC git:(mainline) ✗ NODE_ENV=development gatsby build
GATSBY CONFIG



process.env.NODE_ENV:production
```

**Comment by LekoArts** [comment, marked as answer]
Hi!

We won't change the behavior you're seeing. If you want to use `.env.development` also for production builds you'll need to use another env var to choose, e.g.

```js
// gatsby-config.js

const activeEnv = process.env.ACTIVE_ENV || process.env.NODE_ENV

require('dotenv').config({ path: `.env.${activeEnv}` })
```

And then for a `local:build` script you'd use: `ACTIVE_ENV=development gatsby build`

**Reply by jonmadison** [reply]
Fair enough.

To elaborate on my use case, I don't want to use .env.development for production builds (that makes it sound like i want to do something silly :), I just want to be able to build my elasticsearch index for development. :). According to their docs this (only) happens during "gatsby build"

I'm new to gatsby, so I guess this is something i've learned: "gatsby build", as designed, is intended *only* for a "production" build, and the only way to do an intermediate environment build (i.e. "beta", "gamma", "staging", etc.) would be using the workaround above.

Thanks!

**Reply by Nowaker** [reply]
>If you want to use `.env.development` also for production builds you'll need to use another env var to choose, e.g.

The assumption that `gatsby build` means a "production build" is wrong. There's many situations where `gatsby build` isn't meant for production.

**Reply by jonmadison-amzn** [reply]
as designed, it does (in this case, i'm choosing to disagree and commit. :)  We certainly need to run gatsby build to debug issues that only crop up with the static sites, as well as to deploy to say a pre-prod environment in a static fashion.

**Comment by ankurparihar** [comment]
How'll this work in this case https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-wordpress/docs/plugin-options.md#develophardcachemediafiles

I want to cache wordpress-calls (since I have ~30k pages), and everytime I build again and again I have to fetch those many pages.

How difficult was for you guys to do just

```js
process.env.NODE_ENV = process.env.NODE_ENV || "production"
```

imo, just another dump decision enforced to users by the framework. Gatsby never was meant for large scale sites.
